### ISSUE TITLE
Dictionary is being recognized as void type

### ISSUE THREAD

**Issue by 0verse** [SerenityOS/jakt#493]
This code caught a bug:
```
struct Foo {
    str: String
    num: i64

    function obj(this, mut dict: [String:i64]) throws -> [String:i64] {
        return dict.set(.str, .num)
    }
}
```
The error message I get is this:

```
Error: Cannot call mutating method on an immutable object instance
----- hello.jakt:11:16
 9  |     function obj(this, mutable dict: [String:i64]) throws -> [String:i64] { 
 10 |         return dict.set(.str, .num) 
                  ^- Cannot call mutating method on an immutable object instance
 11 |     } 
-----
```

**Comment by thatbakamono** [comment]
This happens because the type checker doesn't actually check what was modified, it only sees the fact that something was modified and `this` isn't mutable. For reference (callee might be anything, it doesn't have to be a member of `this`):
https://github.com/SerenityOS/jakt/blob/6c01dc6e4d7f25b847d48e4c7f1bb91ba7bfacb5/src/typechecker.rs#L6251-L6257

**Comment by thatbakamono** [comment]
That change unfortunately isn't enough, it quickly backfires in static methods. I checked if it's going to work if I use `current_param_is_mutable` in non static methods and the result of `parse_variable_declaration` in static methods, It seems to be working correctly, doesn't seem like a proper fix tho.

**Comment by cybergsus** [comment]
`current_param_is_mutable` is needed in non-static methods for `mutable this`. So probably a `current_param_is_mutable || var_decl.mutable` would work.

**Comment by cybergsus** [comment]
the tyepchecker is correct, `dict.set` is void (check runtime.jakt)

**Comment by cybergsus** [comment]
Looks like the original issue has been fixed by another commit.

**Comment by 0verse** [comment]
The question is how one is going to set a dictionary inside a function then? Without the return type the sample code is invalid cpp code.

**Comment by lanmonster** [comment]
fixed as of 1ea81c6
